### PR TITLE
Remove usage of max_local_storage_nodes in test infrastructure

### PR DESCRIPTION
**Pull request by ywelsch**
Moves the test infrastructure away from using `node.max_local_storage_nodes`, allowing us in a follow-up PR to deprecate this setting in 7.x and to remove it in 8.0.

This also changes the behavior of InternalTestCluster so that starting up nodes will not automatically reuse data folders of previously stopped nodes. If this behavior is desired, it needs to be explicitly done by passing the data path from the stopped node to the new node that is started.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
I left a couple of minor comments, but this looks good. CI failed without giving any useful feedback:

```
org.elasticsearch.gradle.testclusters.TestClustersPluginIT > testMultiProject FAILED
org.gradle.testkit.runner.UnexpectedBuildFailure at TestClustersPluginIT.java:88
```

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 1]
```diff
@@ -625,11 +616,24 @@ private Settings getNodeSettings(final int nodeId, final long seed, final Settin
...
+        updatedSettings.put(Environment.PATH_HOME_SETTING.getKey(), baseDir);
+
+        if (numDataPaths > 1) {
+            updatedSettings.putList(Environment.PATH_DATA_SETTING.getKey(), IntStream.range(0, numDataPaths).mapToObj(i ->
+                baseDir.resolve(name).resolve("d" + i).toString()).collect(Collectors.toList()));
+        } else {
+            updatedSettings.put(Environment.PATH_DATA_SETTING.getKey(), baseDir.resolve(name));
+        }
+
+        updatedSettings.put(Environment.PATH_SHARED_DATA_SETTING.getKey(), baseDir.resolve(name + "-shared"));
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: yes, it's used by `ESIntegTestCase` which 30% of the time uses  custom data path (see `ESIntegTestCase#indexSettings()`)

File: server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java [hunk 2]
```diff
@@ -88,20 +101,20 @@ public void testRepurpose() throws Exception {
...
+        internalCluster().startDataOnlyNode(dataNodeDataPathSettings);
 
         ensureGreen();
 
         // index is gone.
         assertFalse(indexExists(indexName));
     }
 
-    private void executeRepurposeCommandForOrdinal(Settings settings, String indexUUID, int ordinal,
+    private void executeRepurposeCommandForOrdinal(Settings settings, String indexUUID,
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/env/NodeRepurposeCommandIT.java]: fixed in c10b9a4

**Comment by ywelsch** [comment]
The above test failures are not reproducible after merging in latest master.

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1 
@elasticmachine run elasticsearch-ci/default-distro

**Review by DaveCTurner — APPROVED** [review]
LGTM

(I'm not excluding the possibility that some other tests need adjustment too, but CI didn't find them here)

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**